keps/sig-testing/20190605-presubmitconfig-in-tested-repo.md

```diff
@@ -0,0 +1,259 @@
+      is edge-driven, meaning whenever there is a configuration change
+      it will use the delta to find presubmits that got removed, then
+      check on all pull requests if they contain a status context for
+      a removed presubmit. To make this work for `prow.yaml`, we would
+      have to extend the `status-reconciler` to get an event from GitHub
+      whenever there is a pushevent to the base branch of any open
+      pull request, so that it can find out if a presubmit defined in
+      `prow.yaml` got removed
+
+
+*Alternative:* This could be put into the `prow/plugins/trigger` package
+ as that is the only consumer for this functionality
+
+#### Extend `prow/config` with configuration to opt-in to this feature
+
+This feature should be opt-in to avoid surprises. To achieve that, the
+`ProwConfig` struct in the `prow/config` package has to be extend with
+a `InRepoConfig     map[string]InRepoConfig` property. Also, a
+`func (pc *ProwConfig) InRepoConfigFor(org, repo string) InRepoConfig`
+has to be added that can be used to determine if this functionality is
+enabled. Initially, the `InRepoConfig` struct will be defined as follows:
+
+```
+type InRepoConfig struct {
+  Enabled bool `json:"enabled,omitempty"`
+}
+```
+
+This allows to add further configuration options to this feature in the
+future if needed.
+
+#### Add code to `tide` and the `trigger` plugin so they will use presubmits from a `prow.yaml` if the feature is activated
+
+Both `tide` and the `trigger` plugin will need some code that checks
+if the `prow.yaml` feature is activated and if yes, amend the presubmits
+from it to the presubmits that are configured in the `infra-config` repo.
+
+In `tide` this is very simple, we can just use the `prow/inrepoconfig/api` package to figure out the additional presubmits.
+
+*Open question:* What happens if the merged `prow.yaml` becomes invalid?
+Is tide able to recover from that?
+
+In the `trigger` plugin, we have to check if the `prow.yaml` feature is
+enabled and if yes, use the functionality described from the new `prow/inrepoconfig`
+package to fetch additional presubmits from it.
+
+
+### Risks and Mitigations
+
+#### Plugins may have an inconsitent view on what presubmits exist for a repo
+
+Currently, the presubmit config is exported as a `map[string][]Presubmit`. With the approach outlined above, plugins have to opt-in
+to see presubmits defined in `prow.yaml`.
+
+A possible mitigation for this could be to unexport the `map[string][]presubmit`
+and replace it with a `func (c* Config) Presubmits(org, repo, baseRef string, headRefs []string) ([]Presubmit, error)`.
+This would require adjusments in all plugins that use the presubmit config.
+
+*Open question with this approach:* What do we do about plugins that
+require the presubmit config but do not have a revision they work on,
+like e.G `branchprotector`?
+
+#### Someone could accidentally delete all presubmits in `prow.yaml` and have their PR instantly merged
```

Comment:
Someone could intentionally change prow.yaml to run their bitcoin miner in the Prow cluster as a presubmit on their PR
Someone could intentionally change prow.yaml to run N jobs eating all possible disk space to DDOS the prow cluster

Reply:
> Someone could intentionally change prow.yaml to run their bitcoin miner in the Prow cluster as a presubmit on their PR

That can already be done today by just changing the scripts that are executed by the job.

> Someone could intentionally change prow.yaml to run N jobs eating all possible disk space to DDOS the prow cluster

For a single node, that can already be done today.